README
NAME
    POE::Component::IRC::Plugin::MegaHAL - A PoCo-IRC plugin which provides
    access to a MegaHAL conversation simulator.

SYNOPSIS
    To quickly get an IRC bot with this plugin up and running, you can use
    App::Pocoirc:

     $ pocoirc -s irc.perl.org -j '#bots' -a MegaHAL

    Or use it in your code:

     use POE::Component::IRC::Plugin::MegaHAL;
 
     $irc->plugin_add('MegaHAL', POE::Component::IRC::Plugin::MegaHAL->new(
         Own_channel    => '#bot_chan',
         English        => 1,
         Ignore_regexes => [ qr{^\s*\w+://\S+\s*$} ], # ignore URL-only lines
     ));

DESCRIPTION
    POE::Component::IRC::Plugin::MegaHAL is a POE::Component::IRC plugin. It
    provides "intelligence" through the use of POE::Component::AI::MegaHAL.
    It will talk back when addressed by channel members (and possibly in
    other situations, see "new"). An example:

     --> megahal_bot joins #channel
     <Someone> oh hi there
     <Other> hello there
     <Someone> megahal_bot: hi
     <megahal_bot> oh hi there

    All NOTICEs are ignored, so if your other bots only issue NOTICEs like
    they should, they will be ignored automatically.

    Before using, you should read the documentation for
    POE::Component::AI::MegaHAL and by extension, AI::MegaHAL, so you have
    an idea of what to pass as the 'MegaHAL_args' parameter to "new".

    This plugin requires the IRC component to be POE::Component::IRC::State
    or a subclass thereof.

METHODS
  "new"
    Takes the following optional arguments:

    'MegaHAL', a reference to an existing POE::Component::AI::MegaHAL object
    you have lying around. Useful if you want to use it with multiple IRC
    components. If this argument is not provided, the plugin will construct
    its own object.

    'MegaHAL_args', a hash reference containing arguments to pass to the
    constructor of a new POE::Component::AI::MegaHAL object.

    'Channels', an array reference of channel names. If this is provided,
    the bot will only listen/respond in the specified channels, rather than
    all channels.

    'Own_channel', a channel where it will reply to all messages, as well as
    greet everyone who joins. The plugin will take care of joining the
    channel. It will part from it when the plugin is removed from the
    pipeline. Defaults to none.

    'Abuse_interval', default is 60 (seconds), which means that user X in
    channel Y has to wait that long before addressing the bot in the same
    channel if he wants to get a reply. Setting this to 0 effectively turns
    off abuse protection.

    'Talkative', when set to a true value, the bot will respond whenever
    someone mentions its name (in a PRIVMSG or CTCP ACTION (/me)). If false,
    it will only respond when addressed directly with a PRIVMSG. Default is
    false.

    'Ignore_masks', an array reference of IRC masks (e.g. "purl!*@*") to
    ignore.

    'Ignore_regexes', an array reference of regex objects. If a message
    matches any of them, it will be ignored. Handy for ignoring messages
    with URLs in them.

    'Method', how you want messages to be delivered. Valid options are
    'notice' (the default) and 'privmsg'.

    'English', when set to a true value, some English-language corrections
    will be applied to the bot's output. Currently it will capitalizes the
    word 'I' and make sure paragraphs end with '.' where appropriate.
    Defaults to false.

    Returns a plugin object suitable for feeding to POE::Component::IRC's
    plugin_add() method.

  "brain"
    Takes no arguments. Returns the underlying POE::Component::AI::MegaHAL
    object being used by the plugin.

  "transplant"
    Replaces the brain with the supplied POE::Component::AI::MegaHAL
    instance. Shuts down the old brain if it was instantiated by the plugin
    itself.

AUTHOR
    Hinrik Örn Sigurðsson, hinrik.sig@gmail.com

KUDOS
    Those go to Chris "BinGOs" Williams and his friend GumbyBRAIN.



Changes
0.45  Sun May 15 14:36:09 GMT 2011
    - Fix --version, it was always returning "dev-git"

0.44  Fri May 13 18:28:02 GMT 2011
    - Allow the bot to respond with a CTCP ACTION (/me)

0.43  Tue Sep 28 08:53:59 GMT 2010
    - Update SYNOPSIS to reflect App-Pocoirc changes
    - Fix outdated $VERSION in bin/irchal-seed

0.42  Thu Sep 9 06:29:48 GMT 2010
    - Mention App::Pocoirc in the SYNOPSIS

0.41  Mon Sep 6 04:36:34 GMT 2010
    - Add a 'Channels' parameter

0.40  Wed Sep 1 04:34:40 GMT 2010
    - Fix erroneous dependency (s/Hailo/AI::MegaHAL/)

0.39  Tue Aug 17 22:54:48 GMT 2010
    - Use Dist::Zilla
    - Make all warnings fatal

0.38  Sun Mar 14 10:17:48 GMT 2010
    - Prevent double encoding of non-ASCII channel names
    - Support non-ASCII 'Own_channel' parameter

0.37  Wed Feb 24 02:39:35 GMT 2010
    - Always decode bytes before we use them
    - Fix parsing of pocoirc logs without date stamps

0.36  Fri Feb 19 00:01:26 GMT 2010
    - Read IRC lines as :bytes by default in irchal-seed. This
      avoids issues where PERL_UNICODE or the -C option is set to
      something which tells perl to read STDIN in UTF-8 by default.

0.35  Sun Feb 14 19:01:14 GMT 2010
    - Add META.yml to MANIFEST

0.34  Sun Feb 14 18:50:45 GMT 2010
    - Don't ignore abusers, just don't reply to them

0.33  Fri Feb 12 00:07:24 GMT 2010
    - Print a warning if POE catches an exception in our code
    - Wait for MegaHAL to save the brain before we quit

0.32  Thu Feb 11 21:03:30 GMT 2010
    - When we're addressed, trim some whitespace after the nickname
    - Make MegaHAL.pm and irchal-seed use the same regex to identify lines
      where the bot is addressed
    - irchal-seed: Remove Encode code in favor of irc_to_utf8 from pocoirc

0.31  Thu Jan 21 21:39:39 GMT 2010
    - irchal-seed: Amend bot-nick-catching regex a bit
    - irchal-seed: Document the --help and --version options
    - Converted distribution over to Module::Install
    - Test irchal-seed with Test::Script

0.30  Tue Jan 19 23:49:23 GMT 2010
    - Fix typo in subroutine name

0.29  Tue Jan 19 23:42:09 GMT 2010
    - Apply English language corrections to initial_greeting too

0.28  Tue Jan 19 23:18:31 GMT 2010
    - irchal-seed: Prepend a space to lines beginning with '#' so MegaHAL
      won't ignore them in megahal.trn
    - irchal-seed: Add a --version option
    - Forgot to decode data from MegaHAL's initial_greeting
    - Depend on POE::Component::IRC 6.20 for some bug fixes

0.27  Sat Nov 14 14:22:06 GMT 2009
    - Fix a problem with the English correction, especially when non-ASCII
      characters are involved

0.26  Fri Nov 13 18:32:41 GMT 2009
    - Various documentation enhancements

0.25  Fri Nov 13 15:28:54 UTC 2009
    - Add English punctuation correction feature

0.24  Fri May  8 11:56:34 UTC 2009
    - Remove some duplicate code which was causing an warning

0.23  Tue May  5 15:03:41 UTC 2009
    - irchal-seed: Always strip color and formatting codes
    - Depend on new PoCo::AI::MegaHAL which has a bugfix

0.22  Tue May  5 13:58:11 GMT 2009
    - irchal-seed: Always output text in UTF-8

0.21  Thu Apr 30 12:58:33 GMT 2009
    - Updated documentation
    - Outsource recoding to irc_to_utf8() of PoCo::IRC 6.06

0.20  Fri Mar 27 19:04:53 GMT 2009
    - Fix error in Makefile.PL

0.19  Wed Mar 25 23:49:54 GMT 2009
    - Forgot to import some functions. Oops.

0.18  Wed Mar 25 23:45:28 GMT 2009
    - Rename irchal_seed to irchal-seed
    - Enable the stripping of color/formatting codes with --strip

0.17  Wed Mar 25 16:48:34 UTC 2009
    - Add irssi support to irchal_seed
    - Make irchal_seed's casemapping configurable
    - Improve irchal_seed's documentation

0.16  Sun Mar  8 01:59:32 GMT 2009
    - Fix typo which was causing CTCP ACTIONs to be ignored
    - Some Pod updates

0.15  Sat Nov 29 19:28:53 GMT 2008
    - Added a tip on ignoring URL-only lines to the SYNOPSIS

0.14  Sat Nov 29 19:16:34 GMT 2008
    - Added program irchal_seed to create a MegaHAL brain from IRC logs

0.13  Sat Nov 29 15:44:05 GMT 2008
    - It now sees ACTIONs as "$nick $action"

0.12  Sat Nov 15 02:58:06 GMT 2008
    - Renamed 'Ignore' parameter to 'Ignore_masks'
    - Renamed 'Flood_interval' parameter to 'Abuse_interval'
    - Added 'Ignore_regexes' option
    - Added 'Talkative' option
    - Added some documentation on ignoring other bots

0.11  Thu Sep 25 00:52:35 GMT 2008
    - Convert input all to UTF-8 before processing.

0.10  Mon Sep 22 20:33:18 GMT 2008
    - Fix a serious typo. Sigh.

0.09  Mon Sep 22 20:23:34 GMT 2008
    - It wasn't responding to its name was mentioned. Fixed that.

0.08  Mon Sep 22 17:58:07 GMT 2008
    - Refactor internals, get rid of a bug caused by previous bugfix
    - Make flood protection affect the private channel as well

0.07  Mon Sep 22 16:36:27 GMT 2008
    - Fixed bug which caused it to learn things twice

0.06  Sat Aug  9 06:30:32 GMT 2008
    - Strip color/formatting from text before giving it to MegaHAL
    - Greet people who join our own channel

0.05  Tue Aug  5 19:31:42 GMT 2008
    - Don't respond to every CTCP ACTION (except in own channel)

0.04  Tue Aug  5 08:55:38 GMT 2008
    - Fixed the Test::Pod check in 99_pod.t

0.03  Tue Aug  5 06:18:42 GMT 2008
    - Removed some duplicate code
    - Added some error checking to transplant method

0.02  Tue Aug  5 02:13:14 GMT 2008
    - Minor pod updates
    - Actually part own channel on plugin removal

0.01  Tue Aug  5 01:03:12 GMT 2008
    - First CPAN release


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'A PoCo-IRC plugin which provides access to a MegaHAL conversation simulator.',
  'AUTHOR' => 'Hinrik Örn Sigurðsson <hinrik.sig@gmail.com>',
  'BUILD_REQUIRES' => {
    'Test::Script' => '1.07'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'POE-Component-IRC-Plugin-MegaHAL',
  'EXE_FILES' => [
    'bin/irchal-seed'
  ],
  'LICENSE' => 'perl',
  'NAME' => 'POE::Component::IRC::Plugin::MegaHAL',
  'PREREQ_PM' => {
    'IRC::Utils' => '0.05',
    'POE::Component::AI::MegaHAL' => '0',
    'POE::Component::IRC' => '6.22'
  },
  'VERSION' => '0.45',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/POE/Component/IRC/Plugin/MegaHAL.pm
package POE::Component::IRC::Plugin::MegaHAL;
BEGIN {
  $POE::Component::IRC::Plugin::MegaHAL::AUTHORITY = 'cpan:HINRIK';
}
BEGIN {
  $POE::Component::IRC::Plugin::MegaHAL::VERSION = '0.45';
}

use strict;
use warnings FATAL => 'all';
use Carp;
use Encode qw(decode_utf8 encode_utf8 is_utf8);
use IRC::Utils qw(lc_irc matches_mask_array decode_irc strip_color strip_formatting);
use List::Util qw(first);
use POE;
use POE::Component::AI::MegaHAL;
use POE::Component::IRC::Plugin qw(PCI_EAT_NONE);

sub new {
    my ($package, %args) = @_;
    my $self = bless \%args, $package;

    if (ref $self->{MegaHAL} eq 'POE::Component::AI::MegaHAL') {
        $self->{keep_alive} = 1;
    }
    else {
        $self->{MegaHAL} = POE::Component::AI::MegaHAL->spawn(
            ($self->{MegaHAL_args} ? %{ $self->{MegaHAL_args} } : () ),
        );
    }

    $self->{Method} = 'notice' if !defined $self->{Method} || $self->{Method} !~ /privmsg|notice/;
    $self->{abusers} = { };
    $self->{Abuse_interval} = 60 if !defined $self->{Abuse_interval};

    return $self;
}

sub PCI_register {
    my ($self, $irc) = @_;

    if (!$irc->isa('POE::Component::IRC::State')) {
        die __PACKAGE__ . " requires PoCo::IRC::State or a subclass thereof\n";
    }

    $self->{irc} = $irc;
    POE::Session->create(
        object_states => [
            $self => [qw(
                _start
                _sig_DIE
                _save
                _megahal_reply
                _megahal_greeting
                _megahal_saved
                _greet_handler
                _msg_handler
            )],
        ],
    );

    $irc->plugin_register($self, 'SERVER', qw(isupport ctcp_action join public));
    return 1;
}

sub PCI_unregister {
    my ($self, $irc) = @_;

    $irc->yield(part => $self->{Own_channel}) if $self->{Own_channel};
    delete $self->{irc};
    $poe_kernel->post($self->{session_id}, '_save');

    return 1;
}

sub _save {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

    $kernel->post(
        $self->{MegaHAL}->session_id(),
        '_cleanup',
        { event => '_megahal_saved' },
    );
    return;
}

sub _start {
    my ($kernel, $self, $session) = @_[KERNEL, OBJECT, SESSION];
    $kernel->sig(DIE => '_sig_DIE');
    $self->{session_id} = $session->ID();
    $kernel->refcount_increment($self->{session_id}, __PACKAGE__);
    return;
}

sub _sig_DIE {
    my ($kernel, $self, $ex) = @_[KERNEL, OBJECT, ARG1];
    chomp $ex->{error_str};
    warn "Error: Event $ex->{event} in $ex->{dest_session} raised exception:\n";
    warn "  $ex->{error_str}\n";
    $kernel->sig_handled();
    return;
}

sub _megahal_reply {
    my ($self, $info) = @_[OBJECT, ARG0];
    my $reply = $self->_normalize_megahal($info->{reply});
    $reply = encode_utf8($reply);

    if ($reply =~ s/^\x01 //) {
        $self->{irc}->yield('ctcp', $info->{_target}, "ACTION $reply");
    }
    else {
        $self->{irc}->yield($self->{Method}, $info->{_target}, $reply);
    }
    return;
}

sub _megahal_greeting {
    my ($self, $info) = @_[OBJECT, ARG0];
    my $reply = $self->_normalize_megahal($info->{reply});
    $reply = encode_utf8($reply);

    if ($reply =~ s/^\x01 //) {
        $self->{irc}->yield('ctcp', $info->{_target}, "ACTION $reply");
    }
    else {
        $reply = "$info->{_nick}: $reply";
        $self->{irc}->yield($self->{Method}, $info->{_target}, $reply);
    }
    return;
}

sub _megahal_saved {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

    if (!$self->{keep_alive}) {
        $poe_kernel->post($self->{MegaHAL}->session_id(), 'shutdown');
    }
    delete $self->{MegaHAL};
    $poe_kernel->refcount_decrement($self->{session_id}, __PACKAGE__);
    return;
}

sub _ignoring_channel {
    my ($self, $chan) = @_;

    return if $self->{Own_channel} && $self->_is_own_channel($chan);

    if ($self->{Channels}) {
        return 1 if !first {
            my $c = $chan;
            $c = decode_irc($c) if is_utf8($_);
            $_ eq $c
        } @{ $self->{Channels} };
    }
    return;
}

sub _ignoring_user {
    my ($self, $user) = @_;

    if ($self->{Ignore_masks}) {
        my $mapping = $self->{irc}->isupport('CASEMAPPING');
        return 1 if keys %{ matches_mask_array($self->{Ignore_masks}, [$user], $mapping) };
    }

    return;
}

sub _ignoring_abuser {
    my ($self, $user, $chan) = @_;

    # abuse protection
    my $key = "$user $chan";
    my $last_time = delete $self->{abusers}->{$key};
    $self->{abusers}->{$key} = time;

    return 1 if $last_time && (time - $last_time < $self->{Abuse_interval});
    return;
}

sub _msg_handler {
    my ($self, $kernel, $type, $user, $chan, $what) = @_[OBJECT, KERNEL, ARG0..$#_];
    my $nick = $self->{irc}->nick_name();

    return if $self->_ignoring_channel($chan);
    return if $self->_ignoring_user($user);
    $what = _normalize_irc($what);

    # should we reply?
    my $event = '_no_reply';
    if ($self->{Own_channel} && $self->_is_own_channel($chan)
        || $type eq 'public' && $what =~ s/^\s*\Q$nick\E[:,;.!?~]?\s//i
        || $self->{Talkative} && $what =~ /\Q$nick/i)
    {
        $event = '_megahal_reply';
    }

    if ($event eq '_megahal_reply' && $self->_ignoring_abuser($user, $chan)) {
        $event = '_no_reply';
    }

    if ($self->{Ignore_regexes}) {
        for my $regex (@{ $self->{Ignore_regexes} }) {
            return if $what =~ $regex;
        }
    }

    $kernel->post($self->{MegaHAL}->session_id() => do_reply => {
        event   => $event,
        text    => $what,
        _target => $chan,
    });

    return;
}

sub _is_own_channel {
    my $self = shift;
    my $chan = lc_irc(shift);
    my $own  = lc_irc($self->{Own_channel});

    $chan = decode_irc($chan) if is_utf8($own);
    return 1 if $chan eq $own;
    return;
}

sub _greet_handler {
    my ($self, $kernel, $user, $chan) = @_[OBJECT, KERNEL, ARG0, ARG1];

    return if $self->_ignoring_user($user, $chan);
    return if !$self->{Own_channel} || !$self->_is_own_channel($chan);

    $kernel->post($self->{MegaHAL}->session_id() => initial_greeting => {
        event   => '_megahal_greeting',
        _target => $chan,
        _nick   => (split /!/, $user)[0],
    });

    return;
}

sub _normalize_megahal {
    my ($self, $line) = @_;

    $line = decode_utf8($line);
    if ($self->{English}) {
        $line =~ s{\bi\b}{I}g;
        $line =~ s{(?<=\w)$}{.};
    }
    return $line;
}

sub _normalize_irc {
    my ($line) = @_;

    $line = decode_irc($line);
    $line = strip_color($line);
    $line = strip_formatting($line);
    return $line;
}

sub brain {
    my ($self) = @_;
    return $self->{MegaHAL};
}

sub transplant {
    my ($self, $brain) = @_;

    if (ref $brain ne 'POE::Component::AI::MegaHAL') {
        croak 'Argument must be a POE::Component::AI::MegaHAL instance';
    }

    my $old_brain = $self->{MegaHAL};
    $poe_kernel->post($self->{MegaHAL}->session_id(), 'shutdown') if !$self->{keep_alive};
    $self->{MegaHAL} = $brain;
    $self->{keep_alive} = 1;
    return $old_brain;
}

sub S_isupport {
    my ($self, $irc) = splice @_, 0, 2;
    $irc->yield(join => $self->{Own_channel}) if $self->{Own_channel};
    return PCI_EAT_NONE;
}

sub S_ctcp_action {
    my ($self, $irc) = splice @_, 0, 2;
    my $user         = ${ $_[0] };
    my $chan         = ${ $_[1] }->[0];
    my $what         = ${ $_[2] };
    my $chantypes    = join('', @{ $irc->isupport('CHANTYPES') || ['#', '&']});

    return PCI_EAT_NONE if $chan !~ /^[$chantypes]/;

    $poe_kernel->post(
        $self->{session_id},
        '_msg_handler',
        'action',
        $user,
        $chan,
        "\x01 $what",
    );

    return PCI_EAT_NONE;
}

sub S_public {
    my ($self, $irc) = splice @_, 0, 2;
    my $user         = ${ $_[0] };
    my $chan         = ${ $_[1] }->[0];
    my $what         = ${ $_[2] };

    $poe_kernel->post($self->{session_id} => _msg_handler => 'public', $user, $chan, $what);
    return PCI_EAT_NONE;
}

sub S_join {
    my ($self, $irc) = splice @_, 0, 2;
    my $user         = ${ $_[0] };
    my $chan         = ${ $_[1] };

    return PCI_EAT_NONE if (split /!/, $user)[0] eq $irc->nick_name();
    $poe_kernel->post($self->{session_id} => _greet_handler => $user, $chan);
    return PCI_EAT_NONE;
}

1;

=encoding utf8

=head1 NAME

POE::Component::IRC::Plugin::MegaHAL - A PoCo-IRC plugin which provides access to a MegaHAL conversation simulator.

=head1 SYNOPSIS

To quickly get an IRC bot with this plugin up and running, you can use
L<App::Pocoirc|App::Pocoirc>:

 $ pocoirc -s irc.perl.org -j '#bots' -a MegaHAL

Or use it in your code:

 use POE::Component::IRC::Plugin::MegaHAL;
 
 $irc->plugin_add('MegaHAL', POE::Component::IRC::Plugin::MegaHAL->new(
     Own_channel    => '#bot_chan',
     English        => 1,
     Ignore_regexes => [ qr{^\s*\w+://\S+\s*$} ], # ignore URL-only lines
 ));
 
=head1 DESCRIPTION

POE::Component::IRC::Plugin::MegaHAL is a
L<POE::Component::IRC|POE::Component::IRC> plugin. It provides "intelligence"
through the use of L<POE::Component::AI::MegaHAL|POE::Component::AI::MegaHal>.
It will talk back when addressed by channel members (and possibly in other
situations, see L<C<new>|/"new">). An example:

 --> megahal_bot joins #channel
 <Someone> oh hi there
 <Other> hello there
 <Someone> megahal_bot: hi
 <megahal_bot> oh hi there

It will occasionally send CTCP ACTIONS (/me) too, if the reply in question
happens to be based on an earlier CTCP ACTION from someone.

All NOTICEs are ignored, so if your other bots only issue NOTICEs like
they should, they will be ignored automatically.

Before using, you should read the documentation for
L<POE::Component::AI::MegaHAL|POE::Component::AI::MegaHAL> and by extension,
L<AI::MegaHAL|AI::MegaHAL>, so you have an idea of what to pass as the
B<'MegaHAL_args'> parameter to L<C<new>|/"new">.

This plugin requires the IRC component to be
L<POE::Component::IRC::State|POE::Component::IRC::State> or a subclass thereof.

=head1 METHODS

=head2 C<new>

Takes the following optional arguments:

B<'MegaHAL'>, a reference to an existing
L<POE::Component::AI::MegaHAL|POE::Component::AI::MegaHAL> object you have
lying around. Useful if you want to use it with multiple IRC components.
If this argument is not provided, the plugin will construct its own object.

B<'MegaHAL_args'>, a hash reference containing arguments to pass to the
constructor of a new L<POE::Component::AI::MegaHAL|POE::Component::AI::MegaHAL>
object.

B<'Channels'>, an array reference of channel names. If this is provided, the
bot will only listen/respond in the specified channels, rather than all
channels.

B<'Own_channel'>, a channel where it will reply to all messages, as well as
greet everyone who joins. The plugin will take care of joining the channel.
It will part from it when the plugin is removed from the pipeline. Defaults
to none.

B<'Abuse_interval'>, default is 60 (seconds), which means that user X in
channel Y has to wait that long before addressing the bot in the same channel
if he wants to get a reply. Setting this to 0 effectively turns off abuse
protection.

B<'Talkative'>, when set to a true value, the bot will respond whenever
someone mentions its name (in a PRIVMSG or CTCP ACTION (/me)). If false, it
will only respond when addressed directly with a PRIVMSG. Default is false.

B<'Ignore_masks'>, an array reference of IRC masks (e.g. "purl!*@*") to
ignore.

B<'Ignore_regexes'>, an array reference of regex objects. If a message
matches any of them, it will be ignored. Handy for ignoring messages with
URLs in them.

B<'Method'>, how you want messages to be delivered. Valid options are
'notice' (the default) and 'privmsg'.

B<'English'>, when set to a true value, some English-language corrections
will be applied to the bot's output. Currently it will capitalizes the word
'I' and make sure paragraphs end with '.' where appropriate. Defaults to
false.

Returns a plugin object suitable for feeding to
L<POE::Component::IRC|POE::Component::IRC>'s plugin_add() method.

=head2 C<brain>

Takes no arguments. Returns the underlying
L<POE::Component::AI::MegaHAL|POE::Component::AI::MegaHAL> object being used
by the plugin.

=head2 C<transplant>

Replaces the brain with the supplied
L<POE::Component::AI::MegaHAL|POE::Component::AI::MegaHAL> instance. Shuts
down the old brain if it was instantiated by the plugin itself.

=head1 AUTHOR

Hinrik E<Ouml>rn SigurE<eth>sson, hinrik.sig@gmail.com

=head1 KUDOS

Those go to Chris C<BinGOs> Williams and his friend GumbyBRAIN.

=cut
